sstables/consumer.hh

```diff
@@ -311,29 +372,55 @@ public:
             return read_vint_with_len<unsigned_vint>(data, _u64);
         case prestate::READING_SIGNED_VINT_WITH_LEN:
             return read_vint_with_len<signed_vint>(data, _i64);
+        case prestate::READING_UNSIGNED_VINT_LENGTH_BYTES_WITH_LEN_CONTIGUOUS: {
+            const auto n = std::min(_read_bytes_len - _pos, data.size());
+            std::copy_n(data.begin(), n, _read_bytes.front().get_write() + _pos);
+            data.trim_front(n);
+            _pos += n;
+            if (_pos == _read_bytes_len) {
+                _u64 = unsigned_vint::deserialize(
+                        bytes_view(reinterpret_cast<bytes::value_type*>(_read_bytes.front().get_write()), _read_bytes_len));
+                if (read_bytes_contiguous(data, _u64, *_read_bytes_where_contiguous) == read_status::ready) {
+                    _prestate = prestate::NONE;
+                    return read_status::ready;
+                }
+            }
+            break;
+        }
         case prestate::READING_UNSIGNED_VINT_LENGTH_BYTES_WITH_LEN: {
-            const auto n = std::min(_read_bytes.size() - _pos, data.size());
-            std::copy_n(data.begin(), n, _read_bytes.get_write() + _pos);
+            const auto n = std::min(_read_bytes_len - _pos, data.size());
+            std::copy_n(data.begin(), n, _read_bytes.front().get_write() + _pos);
             data.trim_front(n);
             _pos += n;
-            if (_pos == _read_bytes.size()) {
+            if (_pos == _read_bytes_len) {
                 _u64 = unsigned_vint::deserialize(
-                        bytes_view(reinterpret_cast<bytes::value_type*>(_read_bytes.get_write()), _read_bytes.size()));
+                        bytes_view(reinterpret_cast<bytes::value_type*>(_read_bytes.front().get_write()), _read_bytes_len));
                 if (read_bytes(data, _u64, *_read_bytes_where) == read_status::ready) {
                     _prestate = prestate::NONE;
                     return read_status::ready;
                 }
             }
             break;
         }
+        case prestate::READING_BYTES_CONTIGUOUS: {
+            auto n = std::min(_read_bytes_len - _pos, data.size());
+            std::copy(data.begin(), data.begin() + n, _read_bytes.front().get_write() + _pos);
+            data.trim_front(n);
+            _pos += n;
+            if (_pos == _read_bytes_len) {
+                *_read_bytes_where_contiguous = std::move(_read_bytes.front());
+                _prestate = prestate::NONE;
+                return read_status::ready;
+            }
+            break;
+        }
         case prestate::READING_BYTES: {
-            auto n = std::min(_read_bytes.size() - _pos, data.size());
-            std::copy(data.begin(), data.begin() + n,
-                    _read_bytes.get_write() + _pos);
+            auto n = std::min(_read_bytes_len - _pos, data.size());
+            _read_bytes.push_back(make_tracked_temporary_buffer(data.share(0, n), _permit));
```

Comment:
I don't think it's correct to call make_tracked_temporary_buffer() on the result of data.share() here. `data` is already tracked, you you're double-tracking it. @denesb ?

Reply:
I couldn't find the place in code where we're adding tracking to the `data` bufffer, but it makes sense that's it's already tracked. I'm changing it so that we're only adding tracking to the linearizing buffer

Reply:
It's tracked in `tracking_file_impl`.

Reply:
Yes, you are right @avikivity, these buffers should already be tracked, sharing just bumps a reference counter on it.